### PR TITLE
fix: group invite v4

### PR DESCRIPTION
**Pull request by RC047**
# PR Details
I tried this time the `client.acceptGroupV4Invite()` feature is no longer working

## Description
I changed a little from the newest v4 invite function so that the feature of receiving v4 group invitations works again

## Motivation and Context
In some cases this feature may be rarely used, but this greatly affects other features, so this feature also needs to be updated

## How Has This Been Tested
this function was tested and ran successfully like the function from before

## Types of changes

- [ ] Dependency change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly (index.d.ts).

### CODE REVIEW

**Comment by alechkos** [comment]
It seems that i get an InviteV4 object with `fromId` as `undefined`:
```js
{
  inviteCode: 'iNvItEcOdEv4',
  inviteCodeExp: 'timestamp',
  groupId: 'groupId@g.us',
  groupName: 'groupName',
  fromId: undefined, // indefined
  toId: undefined
}
```
which brokes this
```js
let userWid = window.Store.WidFactory.createWid(fromId);
```
with an error thrown:
```js
Evaluation failed: TypeError: e.toJid is not a function
```

Code i used to test the functionality:
```js
client.on('message', async (message) => {
    if (message.type === 'groups_v4_invite') {
        console.log(message.inviteV4);
        await client.acceptGroupV4Invite(message.inviteV4);
});

**Comment by RC047** [comment]
> It seems that i get an InviteV4 object with `fromId` as `undefined`:
> 
> ```js
> {
>   inviteCode: 'iNvItEcOdEv4',
>   inviteCodeExp: 'timestamp',
>   groupId: 'groupId@g.us',
>   groupName: 'groupName',
>   fromId: undefined, // indefined
>   toId: undefined
> }
> ```
> 
> which brokes this
> 
> ```js
> let userWid = window.Store.WidFactory.createWid(fromId);
> ```
> 
> with an error thrown:
> 
> ```js
> Evaluation failed: TypeError: e.toJid is not a function
> ```
> 
> Code i used to test the functionality:
> 
> ```js
> client.on('message', async (message) => {
>     if (message.type === 'groups_v4_invite') {
>         console.log(message.inviteV4);
>         await client.acceptGroupV4Invite(message.inviteV4);
> })
> ```
> 
> But it works perfect when the `fromId` property of inviteV4 object is defined, so what do you think about to pass a message object to the `acceptGroupV4Invite` method and inside it to define the `inviteV4.fromId` with a value of `message.from`?

After I see, there are several differences in the response in `Message`, when the message that just entered the `fromId` and `toId` objects is not found because the response is indeed not in `Message`. If you call the `getQuotedMessage()` function again, the `fromId` and `toId` objects are found! Here are some slight differences in the response object of `Message` and `getQuotedMessage()`. Here's an examples:

# Message
```
{
   type: 'groups_v4_invite',
   from: '6289XXXXXXX@c.us',
   to: '62887XXXXXX@c.us'
}
```
# getQuotedMessage()
```
{
   type: 'groups_v4_invite',
   from: {
      server: 'c.us',
      user: '62891XXXXXX',
      _serialized: '6289XXXXXXX@c.us'
   },
   to: {
      server: 'c.us',
      user: '62887XXXXXX',
      _serialized: '62887XXXXXX@c.us'
    }
}
```

**Comment by alechkos** [comment]
> > After I see, there are several differences in the response in Message, when the message that just entered the fromId and toId objects is not found because the response is undefined in Message. If you call the getQuotedMessage() function, the fromId and toId objects is defined! Here are some slight differences in the response object of Message and getQuotedMessage()

I meant the `fromId` property in `message.inviteV4` object:
```js
client.on('message', async (message) => {
    if (message.type === 'groups_v4_invite') {
        console.log(message.inviteV4.fromId); // output: undefiend
});
```
The object that [`acceptGroupV4Invite`](https://github.com/RC047/whatsapp-web.js/blob/8abf1c0e9fbe4238f1daf6ad9fecb2397202d9a4/src/Client.js#L928C11-L928C30) gets supposed to have `fromId` value defined with the invitor id but instead it is `undefined`,
because of that the [`userWid`](https://github.com/RC047/whatsapp-web.js/blob/8abf1c0e9fbe4238f1daf6ad9fecb2397202d9a4/src/Client.js#L933) is also undefined which then throws an error:
```js
Evaluation failed: TypeError: e.toJid is not a function
```

**Comment by RC047** [comment]
> > > After I see, there are several differences in the response in Message, when the message that just entered the fromId and toId objects is not found because the response is undefined in Message. If you call the getQuotedMessage() function, the fromId and toId objects is defined! Here are some slight differences in the response object of Message and getQuotedMessage()
> 
> I meant the `fromId` property in `message.inviteV4` object:
> 
> ```js
> client.on('message', async (message) => {
>     if (message.type === 'groups_v4_invite') {
>         console.log(message.inviteV4.fromId); // output: undefiend
> });
> ```
> 
> The object that [`acceptGroupV4Invite`](https://github.com/RC047/whatsapp-web.js/blob/8abf1c0e9fbe4238f1daf6ad9fecb2397202d9a4/src/Client.js#L928C11-L928C30) gets supposed to have `fromId` value defined with the invitor id but instead it is `undefined`, because of that the [`userWid`](https://github.com/RC047/whatsapp-web.js/blob/8abf1c0e9fbe4238f1daf6ad9fecb2397202d9a4/src/Client.js#L933) is also undefined which then throws an error:
> 
> ```js
> Evaluation failed: TypeError: e.toJid is not a function
> ```

Yes, so that means the object from `inviteV4` is in the source `./src/structures/Message.js`:
```js
/**
   * Group Invite Data
   * @type {object}
   */
this.inviteV4 = data.type === MessageTypes.GROUP_INVITE ? {
      inviteCode: data.inviteCode,
      inviteCodeExp: data.inviteCodeExp,
      groupId: data.inviteGrp,
      groupName: data.inviteGrpName,
      fromId: data.from._serialized, // undefined
      toId: data.to._serialized // undefined
} : undefined;
```

Maybe something has to be changed from one of them?

**Comment by alechkos** [comment]
> Yes, so that means the object from `inviteV4` is in the source `./src/structures/Message.js`

Cool, i didn't think about it
What about to change the `fromId: data.from._serialized` to `fromId: (typeof (data.from) === 'object' && data.from !== null) ? data.from._serialized : data.from`?
```js
this.inviteV4 = data.type === MessageTypes.GROUP_INVITE ? {
            inviteCode: data.inviteCode,
            inviteCodeExp: data.inviteCodeExp,
            groupId: data.inviteGrp,
            groupName: data.inviteGrpName,
            fromId: (typeof (data.from) === 'object' && data.from !== null) ? data.from._serialized : data.from, // here
            toId: data.to._serialized
        } : undefined;
```

This will ensure the inviteV4 that is sent to the `acceptGroupV4Invite()`  wont throw an error

**Comment by RC047** [comment]
> 

Maybe you can, and apply it to the `toId` object as well

**Comment by alechkos** [comment]
> Maybe you can, and apply it to the `toId` object as well

Yes but since you have this pr opened you can simply add another commit that applies those changes, this action will also make sence within your pr.

**Comment by RC047** [comment]
> > Maybe you can, and apply it to the `toId` object as well
> 
> Yes but since you have this pr opened you can simply add another commit that applies those changes, this action will also make sence within your pr.

Sure.

File: src/structures/Message.js [hunk 1]
```diff
@@ -168,8 +168,8 @@ class Message extends Base {
             inviteCodeExp: data.inviteCodeExp,
             groupId: data.inviteGrp,
             groupName: data.inviteGrpName,
-            fromId: data.from._serialized,
-            toId: data.to._serialized
+            fromId: data.from._serialized ? data.from._serialized : data.from,
```
Comment (alechkos) [inline src/structures/Message.js:171]: You will get a TypeError in a case of `data.from` is null or undefined or if `data.from` is a property but not an object

Comment (RC047) [reply src/structures/Message.js]: Yes, I know that, it's just that I'm trying to make the code not too long and simpler

**Review by alechkos — APPROVED** [review]
Works here

**Comment by alechkos** [comment]
closes #2060

**Comment by joweste** [comment]
Hi alechkos. Could you fix setDescription for groups? It's broken. The methods never return from call.

**Comment by RC047** [comment]
> Hi alechkos. Could you fix setDescription for groups? It's broken. The method never return from call.

You can use `await` in `window.Store.MsgKey.newId()` method

**Comment by joweste** [comment]
> > Hi alechkos. Could you fix setDescription for groups? It's broken. The method never return from call.
> 
> You can use `await` in `window.Store.MsgKey.newId()` method

Yes, but when I use, no description is setting. Here is the code. After execute the code, I have checked the groupMetadata desc and it is undefined.
PS: wbot is my 'client' equivalent
      
      ```
const id='....'
      const tchats = await wbot.getChats();
      const ochat = tchats.filter(mchat => {
        return mchat.isGroup && mchat.id._serialized === id;
      });
      const tchat = ochat?.[0] as WAWebJS.GroupChat;
      if (tchat) {
        tchat.setDescription(newDescription);
      } 
```

**Review by shirser121 — APPROVED** [review]
Looking great. Thanks

**Comment by shirser121** [comment]
@tuyuribr3 @pedroslopez @aliyss

**Review by tuyuribr — APPROVED** [review]

**Pull request merged**